### PR TITLE
shortcode

### PR DESCRIPTION
**Pull request by kavehmz**

### CODE REVIEW

File: lib/Finance/Contract.pm [hunk 1]
```diff
@@ -248,6 +248,20 @@ has starts_as_forward_starting => (
     default => 0,
 );
 
+=head2 barrier_pip_size
```
Comment (jy-binary) [inline lib/Finance/Contract.pm]: I think this should be underlying_pip_size, or just pip_size. It's not technically the pip size of the barrier, but of the underlying market of this contract

File: lib/Finance/Contract.pm [hunk 2]
```diff
@@ -248,6 +248,20 @@ has starts_as_forward_starting => (
...
+
+Barrier pip size the minimum fluctuation amount for type of market. It is normally fraction.
+It must be supplier on initialization or packge which is using this package must have a builder to build this value.
```
Comment (jy-binary) [inline lib/Finance/Contract.pm]: remove mention of a builder. I think it needs to be passed in on construction. If anyone wants to do magic with builders, they can do that but it isn't the recommended way

File: lib/Finance/Contract.pm [hunk 3]
```diff
@@ -579,9 +593,9 @@ sub shortcode {
     my @shortcode_elements = ($contract_type, $self->underlying->symbol, $self->payout, $shortcode_date_start, $shortcode_date_expiry);
 
     if ($self->two_barriers) {
-        push @shortcode_elements, map { _barrier_for_shortcode_string($_, $contract_type) } ($self->supplied_high_barrier, $self->supplied_low_barrier);
+        push @shortcode_elements, map { $self->_barrier_for_shortcode_string($_, $contract_type) } ($self->supplied_high_barrier, $self->supplied_low_barrier);
```
Comment (jy-binary) [inline lib/Finance/Contract.pm]: out of interest why did you need to add $self-> ?

File: lib/Finance/Contract.pm [hunk 4]
```diff
@@ -716,13 +730,27 @@ sub _barrier_from_shortcode_string {
 
 # Generates a string version of a barrier by multiplying the actual barrier to remove the decimal point
 sub _barrier_for_shortcode_string {
-    my ($string, $contract_type) = @_;
+    my ($self, $string, $contract_type) = @_;
+
+    # Do not manipulate relatvie barriers.
```
Comment (jy-binary) [inline lib/Finance/Contract.pm]: typo in relatvie

File: lib/Finance/Contract.pm [hunk 5]
```diff
@@ -716,13 +730,27 @@ sub _barrier_from_shortcode_string {
...
+    return $string if not looks_like_number($string);
 
+    $string = $self->_pipsized_value($string);
     $string *= _FOREX_BARRIER_MULTIPLIER if $contract_type !~ /^DIGIT/ and $string and looks_like_number($string);
 
+    # Make sure it's an integer
+    $string = roundnear(1, $string);
```
Comment (basiliscos) [inline lib/Finance/Contract.pm]: please, use `sprintf` for everything related to barriers; we'll probably need the same for C++ code, and `sprintf` is more portalble.

File: lib/Finance/Contract.pm (hunk 3 above)
Comment (kavehmz) [inline lib/Finance/Contract.pm]: To get the pip_size from object.

File: lib/Finance/Contract.pm (hunk 1 above)
Comment (kavehmz) [inline lib/Finance/Contract.pm]: name is changed to pip_size
